File: 1002-find-common-characters/1002-find-common-characters.cs
public class Solution {
    public IList<string> CommonChars(string[] words) {
        if (words.Length == 0) return new List<string>();

        Dictionary<char, int> commonCount = new Dictionary<char, int>();
        foreach (char c in words[0])
        {
            if (commonCount.ContainsKey(c))
            {
                commonCount[c]++;
            }
            else
            {
                commonCount[c] = 1;
            }
        }

        for (int i = 1; i < words.Length; ++i)
        {
            Dictionary<char, int> currentCount = new Dictionary<char, int>();
            foreach (char c in words[i])
            {
                if (currentCount.ContainsKey(c))
                {
                    currentCount[c]++;
                }
                else
                {
                    currentCount[c] = 1;
                }
            }

            List<char> keysToRemove = new List<char>();
            foreach (var pair in commonCount)
            {
                if (currentCount.ContainsKey(pair.Key))
                {
                    commonCount[pair.Key] = Math.Min(pair.Value, currentCount[pair.Key]);
                }
                else
                {
                    keysToRemove.Add(pair.Key);
                }
            }

            foreach (char key in keysToRemove)
            {
                commonCount.Remove(key);
            }
        }

        List<string> result = new List<string>();
        foreach (var pair in commonCount)
        {
            for (int i = 0; i < pair.Value; ++i)
            {
                result.Add(pair.Key.ToString());
            }
        }

        return result;
    }
}

File: 1002-find-common-characters/1002-find-common-characters.cpp
class Solution {
public:
    vector<string> commonChars(vector<string>& words) {
        
       if (words.empty()) return {};

        unordered_map<char, int> commonCount;
        for (char c : words[0]) {
            commonCount[c]++;
        }

        for (auto i = 1; i < words.size(); ++i) {
            unordered_map<char, int> currentCount;
            for (char c : words[i]) {
                currentCount[c]++;
            }

            for (auto& pair : commonCount) {
                pair.second = min(pair.second, currentCount[pair.first]);
            }

            for (auto it = commonCount.begin(); it != commonCount.end();) {
                if (currentCount.find(it->first) == currentCount.end()) {
                    it = commonCount.erase(it);
                } else {
                    ++it;
                }
            }
        }
        
        vector<string> result;
        for (auto& pair : commonCount) {
            for (int i = 0; i < pair.second; ++i) {
                string s = "";
                s+=pair.first;
                result.push_back(s);
            }
        }

        return result;
    }
};